src/main/java/appeng/thirdparty/codechicken/lib/model/pipeline/BakedPipeline.java

```diff
@@ -0,0 +1,434 @@
+ * meaning that the internal states for the transformers can be safely shared across VertexFormats, this reduces
+ * array creations, and generally makes the system as efficient as it is.
+ *
+ * To use the system:
+ * Grab any elements you need to set state data on first, using {@link #getElement(String, Class)}
+ * transformers should NOT clear their state on pipeline Reset's so set any global data on elements now.
+ * Assuming you are looping over a set of quads to transform, next you need to {@link #reset} the pipeline,
+ * Now you should disable / enable any optional elements that are needed, NOTE: Element states are reset when resetting
+ * the pipeline.
+ * Now you will need to call {@link #prepare(IVertexConsumer)} on the pipeline, here you will pass your collector,
+ * usually this is some form of (Unpacked)BakedQuadBuilder, See {@link QuadBuilder} for a simple and fast implementation
+ * for standard BakedQuads, and {@link UnpackedBakedQuad.Builder} for UnpackedBakedQuads.
+ * Now final step, simply pipe the quad you want to transform INTO the pipeline 'quad.pipe(pipeline)'
+ * And that's it! hell, Pipe a pipeline into each other for all i care, the system is efficient enough that there
+ * would be no performance penalty for doing so.
+ *
+ * @author covers1624
+ */
+public class BakedPipeline implements ISmartVertexConsumer
+{
+
+	private PipelineElement[] elements;
+	private Map<String, PipelineElement> nameLookup;
+	private IPipelineConsumer first;
+
+	private Quad unpacker = new Quad();
+
+	private BakedPipeline( PipelineElement[] elements )
+	{
+		this.elements = elements;
+		nameLookup = Arrays.stream( elements ).collect( Collectors.toMap( e -> e.name, e -> e ) );
+	}
+
+	/**
+	 * Used to create a BakedPipeline.
+	 *
+	 * @return The builder.
+	 */
+	public static Builder builder()
+	{
+		return new Builder();
+	}
+
+	/**
+	 * Used to reset the pipeline for the next quad.
+	 * MUST be called between quads.
+	 *
+	 * @param format The format.
+	 */
+	public void reset( VertexFormat format )
+	{
+		reset( CachedFormat.lookup( format ) );
+	}
+
+	/**
+	 * Used to reset the pipeline for the next quad.
+	 * MUST be called between quads.
+	 *
+	 * @param format The format.
+	 */
+	public void reset( CachedFormat format )
+	{
+		unpacker.reset( format );
+		for( PipelineElement element : elements )
+		{
+			element.reset( format );
+		}
+		first = null;
+	}
+
+	/**
+	 * Get an element from the pipeline.
+	 *
+	 * @param name The name of the element.
+	 * @param clazz The Class of the element, used to safe cast.
+	 *
+	 * @return The element.
+	 */
+	public <T extends IPipelineConsumer> T getElement( String name, Class<T> clazz )
+	{
+		PipelineElement element = nameLookup.get( name );
+		if( element != null )
+		{
+			if( !clazz.isAssignableFrom( element.consumer.getClass() ) )
+			{
+				throw new IllegalArgumentException( "Element with name " + name + " is not assignable from reference class." );
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "Element with name " 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1192)